README.md
# Xataface Depselect Module

A dependent select widget for [Xataface](https://github.com/shannah/xataface) forms. A dependent select is one whose options are determined by the state of other widgets in the form. It listens for changes to other selects and updates its own options accordingly.

A common example is a form with select lists for Country, Province, and City. The Province options depend on the selected Country, and the City options depend on the selected Province. The depselect widget handles this by live-updating options whenever a dependency field changes.

## Features

- **Live updating** of options when dependency fields change.
- **Checkbox display mode** for multi-select scenarios (e.g. selecting multiple rooms in a building).
- **"Add" button** to create new options inline via a dialog with a new record form.
- **Permissions integration** with Xataface's permissions system — options are only loaded if the user has `view` permission on the target table (can be overridden).
- **No valuelist required** — options are loaded on demand, which can be more efficient for large tables since valuelists are loaded in their entirety on every request.

## Requirements

- Xataface 1.4 or higher

## Installation

1. Copy the module into your `modules` directory:

   ```
   modules/depselect/
   ```

2. Add the following to the `[_modules]` section of your app's `conf.ini`:

   ```ini
   modules_depselect=modules/depselect/depselect.php
   ```

## Usage

To use the depselect widget, set `widget:type=depselect` in the appropriate `fields.ini` file for your table. The following directives are available:

### Directives

| Directive | Required | Description |
|---|---|---|
| `widget:type` | Yes | Must be set to `depselect`. |
| `widget:table` | Yes | The table from which options are loaded. |
| `widget:filters:xyz` | No | Filters applied when loading options. Can be any valid Xataface GET parameter. Use `$field_name` syntax to reference other form fields as dynamic filter values. |
| `widget:keycol` | No | The column to use as the option value/key. Defaults to the table's primary key. |
| `widget:labelcol` | No | The column to use as the option display label. |
| `widget:displayType` | No | Set to `checkbox` to render options as checkboxes for multi-select. Default is `select` (dropdown). |
| `widget:ignore_permissions` | No | Set to bypass the default requirement that the user has `view` permission on the options table. |

### Display Types

#### `select` (default)

Renders a standard dropdown select list. The user can choose a single value. This is the default behavior when `widget:displayType` is not specified.

#### `checkbox`

Renders options as a group of checkboxes, allowing the user to select multiple values. Selected values are stored as a comma-separated string (e.g. `"1,3,5"`).

This mode is useful when a record needs to be associated with multiple options from a dependent table — for example, selecting multiple rooms within a building.

**Note:** When using `checkbox` mode, the database column should be a `VARCHAR` or `TEXT` field large enough to hold comma-separated IDs rather than a single foreign key integer.

## Examples

### Basic chained dropdowns

A city field dependent on a province field, which is dependent on a country field:

```ini
[country_id]
    widget:type=depselect
    widget:table=countries

[province_id]
    widget:type=depselect
    widget:table=provinces
    widget:filters:country_id="$country_id"

[city_id]
    widget:type=depselect
    widget:table=cities
    widget:filters:province_id="$province_id"
```

### Checkbox multi-select

A work order form where the user selects a branch, then a building, then checks off one or more rooms. The `rooms` field uses `widget:displayType=checkbox` so that multiple rooms can be selected:

```ini
[branch_id]
    widget:type=depselect
    widget:table=branches
    widget:keycol=branch_id
    widget:labelcol=branch_name

[building_component_id]
    widget:type=depselect
    widget:table=building_components
    widget:keycol=building_component_id
    widget:labelcol=code
    widget:filters:branch_id="$branch_id"

[rooms]
    widget:type=depselect
    widget:table=rooms
    widget:keycol=room_id
    widget:labelcol=room_number
    widget:displayType=checkbox
    widget:filters:building_component_id="$building_component_id"
```

When the user selects a building, the available rooms for that building are displayed as checkboxes. The selected room IDs are saved as a comma-separated string (e.g. `"101,103"`) in the `rooms` column.

## Support

- [Xataface Forum](http://xataface.com/forum)

## Credits

- Developed by Steve Hannah (steve@weblite.ca)
- Built with [jQuery](https://jquery.com) and [jQuery UI](https://jqueryui.com)

## License

This library is free software; you can redistribute it and/or modify it under the terms of the GNU Library General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the [GNU Library General Public License](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.html) for more details.


changes.txt
Version 0.1 2345
June 8, 2011

	- Initial check-in

widget.php
<?php
import('Dataface/FormTool/select.php');
/*
 * Xataface DepSelect Module
 * Copyright (C) 2011  Steve Hannah <steve@weblite.ca>
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 * 
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
 * Boston, MA  02110-1301, USA.
 *
 */
/**
 * @brief A Dataface_FormTool wrapper for building depselect widgets in Dataface_QuickForm forms.
 *
 * All widget types require a wrapper of this kind to implement the glue between the field and the 
 * database records.  This particular wrapper only implements the buildWidget() method but
 * it could also implement pushValue() and pullValue() methods to define how data should be treated
 * when passing between Dataface_RecordObjects and the HTML_QuickForm widget.
 *
 * Note that the modules_depselect class actually registers this class with Dataface_FormTool so that
 * it knows of its existence.
 *
 */
class Dataface_FormTool_depselect extends Dataface_FormTool_select {

	/**
	 * Defines how a depselect widget should be built.
	 *
	 * @param Dataface_Record $record The Dataface_Record that is being edited.
	 * @param array &$field The field configuration data structure that the widget is being generated for.
	 * @param HTML_QuickForm The form to which the field is to be added.
	 * @param string $formFieldName The name of the field in the form.
	 * @param boolean $new Whether this widget is being built for a new record form.
	 * @return HTML_QuickForm_element The element that can be added to a form.
	 *
	 */
	function &buildWidget(&$record, &$field, &$form, $formFieldName, $new=false){
		$el = parent::buildWidget($record, $field, $form, $formFieldName, $new);
		$mt = Dataface_ModuleTool::getInstance();
		$mod = $mt->loadModule('modules_depselect');
		$atts = $el->getAttributes();
		if ( !@$atts['class'] ) $atts['class'] = '';
		$atts['class'] .= ' xf-depselect';
		if ( !@$atts['data-xf-table'] ){
			$atts['data-xf-table'] = $field['tablename'];
		}
		$targetTable = Dataface_Table::loadTable($field['widget']['table']);
		if ( PEAR::isError($targetTable) ){
			error_log("Your field $formFieldName is missing the widget:table directive or the table does not exist.");
			throw new Exception("Your field $formFieldName is missing the widget:table directive or the table does not exist.");
		}
		$targetPerms = $targetTable->getPermissions();
		if ( @$targetPerms['new'] ){
			$atts['data-xf-depselect-options-table'] = $field['widget']['table'];
		}
		$atts['df:cloneable'] = 1;
		
		
		
		$jt = Dataface_JavascriptTool::getInstance();
		$jt->addPath(dirname(__FILE__).'/js', $mod->getBaseURL().'/js');
		
		$ct = Dataface_CSSTool::getInstance();
		$ct->addPath(dirname(__FILE__).'/css', $mod->getBaseURL().'/css');
		
		// Add our javascript
		$jt->import('xataface/widgets/depselect.js');
		
		$filters = array();
		if ( @$field['widget']['filters']  and is_array($field['widget']['filters'])){
			foreach ($field['widget']['filters'] as $key=>$val){
				$filters[] = urlencode($key).'='.urlencode($val);
			}
		}
		
		$atts['data-xf-depselect-filters'] = implode('&', $filters);
	
	
		
		if ( @$field['widget']['nomatch'] ){
			$atts['data-xf-depselect-nomatch'] = $field['widget']['nomatch'];
		}
		
		$el->setAttributes($atts);
	
		return $el;
	}
	

}

depselect.php
<?php
/*
 * Xataface Depselect Module
 * Copyright (C) 2011  Steve Hannah <steve@weblite.ca>
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 * 
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
 * Boston, MA  02110-1301, USA.
 *
 */
 
/**
 * @brief The main depselect module class.  This loads all of the dependencies for the
 * 	module.
 *
 * Of note, this module depends on the XataJax module for the loading
 * of its javascripts.
 *
 */
 class modules_depselect {
	/**
	 * @brief The base URL to the depselect module.  This will be correct whether it is in the 
	 * application modules directory or the xataface modules directory.
	 *
	 * @see getBaseURL()
	 */
	private $baseURL = null;
	
	
	
	/**
	 * @brief Initializes the depselect module and registers all of the event listener.
	 *
	 */
	function __construct(){
		$app = Dataface_Application::getInstance();
		
		
		// Now work on our dependencies
		$mt = Dataface_ModuleTool::getInstance();
		
		// We require the XataJax module
		// The XataJax module activates and embeds the Javascript and CSS tools
		$mt->loadModule('modules_XataJax', 'modules/XataJax/XataJax.php');
		
		
		// Register the tagger widget with the form tool so that it responds
		// to widget:type=tagger
		import('Dataface/FormTool.php');
		$ft = Dataface_FormTool::getInstance();
		$ft->registerWidgetHandler('depselect', dirname(__FILE__).'/widget.php', 'Dataface_FormTool_depselect');
		
		
	}
	
	
	/**
	 * @brief Returns the base URL to this module's directory.  Useful for including
	 * Javascripts and CSS.
	 *
	 */
	public function getBaseURL(){
		if ( !isset($this->baseURL) ){
			$this->baseURL = Dataface_ModuleTool::getInstance()->getModuleURL(__FILE__);
		}
		return $this->baseURL;
	}
}

js/xataface/widgets/depselect.js
/*
 * Xataface Depselect Module
 * Copyright (C) 2011  Steve Hannah <steve@weblite.ca>
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 * 
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
 * Boston, MA  02110-1301, USA.
 *
 */
 
 
//require <jquery.packed.js>
//require <jquery-ui.min.js>
//require-css <jquery-ui/jquery-ui.css>
//require <RecordDialog/RecordDialog.js>
//require-css <xataface/widgets/depselect.css>  
(function(){
	var $ = jQuery;
	
	
	/**
	 * Finds a field by name relative to a starting point.  It will search only within
	 * the startNode's form group (i.e. class xf-form-group).
	 *
	 * @param {HTMLElement} startNode The starting point of our search (we search for siblings).
	 * @param {String} fieldName The name of the field we are searching for.
	 *
	 * @return {HTMLElement} The found field or null if it cannot find it.
	 */
	function findField(startNode, fieldName){
		
		var parentGroup = $(startNode).parents('.xf-form-group').get(0);
		if ( !parentGroup ) parentGroup = $(startNode).parents('form').get(0);
		if ( !parentGroup ) return null;
		//alert('here');
		var fld = $('[data-xf-field="'+fieldName+'"]', parentGroup).get(0);
		return fld;
	}
	
	
	/**
	 * Updates the values for a depselect list.  This is usually called in 
	 * response to a change in one of the selects that this depselect
	 * is dependent upon.
	 *
	 * @param {HTMLElement} select The depselect <select> element that is to be updated.
	 * @param {Object} filters The filters that should be applied to the options list.
	 * 		These filters may be as yet unprocessed.  If a value begins with a $, that 
	 * 		means that it should be substituted with the value of that field.
	 *
	 */
	function updateValuesFor(select, filters){
	
		var tablename = $(select).attr("data-xf-table");
		var fieldname = $(select).attr('data-xf-field');
		
		var url = DATAFACE_SITE_HREF;
		var q = {
			'-action': 'depselect_load',
			'-table': tablename,
			'-field': fieldname
		};
		
		$.each(filters, function(key,val){
			if ( !key ) return;
			if ( val.indexOf('$') == 0 ){
				var fname = val.substr(1);
				var field = findField(select, fname);
				if ( field && $(field).val() ){
					q[key] = $(field).val();
				} else {
					q[key] = '=';
				}
			} else {
				
				q[key] = val;
			}
		});
		
		$.get(url, q, function(res){
			try {
				if ( typeof(res) == 'string' ){
					eval('res='+res+';');
				}
				if ( res.code == 200 ){
				
					var currVal = $(select).val();
					select.options.length=1;
					$.each(res.values, function(key,val){
						$(select).append(
							$('<option></option>')
								.attr('value', key)
								.text(val)
						);
					});
					
					$(select).val(currVal);
				} else {
					if ( res.message ) throw res.message;
					else throw 'Failed to load values for field '+fieldname+' because of an unspecified server error.';
				}
				
			
			} catch (e){
				alert(e);
			}
		
		});
	}
	
	
	/**
	 * Adds an option to the given select list.  This uses the record 
	 * dialog to pop up with a "new record form" in an internal dialog.
	 *
	 * @param {HTMLElement} select The select list to add an option to.
	 * @param {Object} filters The filters to apply.
	 */
	function addOptionFor(select, filters){
		var tableName = $(select).attr("data-xf-depselect-options-table");
		if ( !tableName ) return;
		
		var q = {};
		
		$.each(filters, function(key,val){
			if ( !key ) return;
			if ( val.indexOf('$') == 0 ){
				var fname = val.substr(1);
				var field = findField(select, fname);
				if ( field && $(field).val() ){
					q[key] = $(field).val();
				} else {
					//q[key] = '=';
				}
			} else {
				
				q[key] = val;
			}
		});
		
		var dlg = new xataface.RecordDialog({
			table: tableName,
			callback: function(data){
			
				updateValuesFor(select, filters);
				
			},
			params: q
		
		});
		
		dlg.display();
		
	}
	
	
	/**
	 * When defining the javascript for a widget, we always wrap it in
	 * registerXatafaceDecorator so that it will be run whenever any new content is
	 * loaded ino the page.  This makes it compatible with the grid widget.
	 *
	 * If you don't do this, the widget will only be installed on widgets at page load time
	 * so when new rows are added via the grid widget, the necessary javascript won't be installed
	 * on those widgets.
	 */
	registerXatafaceDecorator(function(node){
		
		$('select.xf-depselect', node).each(function(){
			
			var self = this;
			var filtersAttr = $(this).attr('data-xf-depselect-filters');
			var filters = {};
			filtersAttr = filtersAttr.split('&');
			$.each(filtersAttr, function(){
				var parts = this.split('=');
				filters[decodeURIComponent(parts[0])] = decodeURIComponent(parts[1]);
			});
			
			
			// Now that we have our filters we should start listening for changes
			// on each of the filters.
			
			$.each(filters, function(key,val){
			
				if ( val.indexOf('$') == 0 ){
					// It is a variable
					var depField = val.substr(1);
					var field = findField(self, depField);
					//alert(depField);
					if ( !field ) return;
					
					// We want to listen or changes to this field
					// so that we can update our values whenever the field
					// changes.
					$(field).change(function(){
						//alert('value changed');
						updateValuesFor(self, filters);
					});
				}
			
			});
			
			
			if ( $(self).attr("data-xf-depselect-options-table") ){
				// We only add this button if the user has permission 
				// to add new records to the target table.
				$('<a><img src="'+DATAFACE_URL+'/images/add_icon.gif"/></a>')
					.addClass('xf-depselect-add-btn')
					.click(function(){
						addOptionFor(self, filters);
					})
					.insertAfter(self);
			}
			
			
			
			
			
			// Initialize the values to begin with.
			updateValuesFor(self, filters);
			
			
		});
		
		
	
	});
})();

actions/depselect_load.php
<?php
/*
 * Xataface Depselect Module
 * Copyright (C) 2011  Steve Hannah <steve@weblite.ca>
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 * 
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
 * Boston, MA  02110-1301, USA.
 *
 */
 
/**
 * @brief An HTTP REST action to load the options for a particular depselect field.  Returns JSON
 * data.
 *
 * @section rest REST Interface
 *
 * @param string -table The name of the table where the depselect field resides.
 * @param string -field The name of the depselect field.
 * @return JSON A JSON data structure of the following form:
 * @code
 * {
 *     code: <int>,			// The status code.  200 for success
 *     message: <string>, 	// String message.  Useful if there is an error.
 *     values: <Object>		// A JSON object with the options to be returned.
 * }
 * @endcode
 *
 * @subsection statuscodes Status Codes
 *
 * <table>
 *		<tr>
 *			<th>Code</th>
 *			<th>Meaning</th>
 *		</tr>
 *		<tr>
 *			<td>200</th>
 *			<td>Success.  The values should be contained in the 'values' property.</td>
 *		</tr>
 *		<tr>
 *			<td>401</td>
 *			<td>Permission Denied.  The user doesn't have permission to access these values.  In order
 *				to access this action, the user must either have edit or new permission on the 
 *				source table (i.e. the table containing the depselect field).
 *			</td>
 *		</tr>
 *		<tr>
 *			<td>500</td>
 *			<td>Server Error or Configuration Error.  Check the server error log for details.</td>
 *		</tr>
 *	</table>
 *
 * @subsection samplevalues Sample Values Object
 *
 * The "values" property will be an object with just key values pairs, where the key
 * is the id that is meant to be stored in the database.  The value is the label that is
 * displayed to the user.  E.g.
 *
 * @code
 * {
 *     1: 'Option #1',
 *     2: 'Option #2',
 *	.. etc...
 * }
 * @endcode
 */
class actions_depselect_load {


	/**
	 * @brief Returns  the default error message for when a server error has occurred.
	 * @param string $fieldname The name of the depselect field on which the error occurred.
	 * @return string The full error message.
	 */
	static function errorMessage($fieldname){
	
		return "Failed to load values for field '".$fieldname."'.  See server error log for details.";
	}

	/**
	 * @brief Handles the HTTP request.  Outputs the JSON response.
	 */
	function handle($params){
	
		session_write_close();
		header('Connection: close');
		
		$app = Dataface_Application::getInstance();
		$query = $app->getQuery();
		$table = Dataface_Table::loadTable($query['-table']);
		$field = $table->getField($query['-field']);
		//$table = Dataface_Table::loadTable($field['widget']['table']);
		$this->loadFromTable($table, $field, $query);
		
	}
	
	
	/**
	 * @brief Loads the values for a particular field in a particular table.
	 *
	 * This method is delegated to by handle() for cleanliness purposes.
	 *
	 * @param Dataface_Table $table The table where the depselect field resides.
	 * @param array &$field The field definition of the depselect field.
	 * @param array $query The current GET parameters.
	 * 
	 */
	private function loadFromTable(Dataface_Table $table, &$field, $query){
		
		try {
			
			$perms = $table->getPermissions(array('field'=>$field['name']));
			if ( !@$perms['edit'] and !@$perms['new'] ){
				// The user doesn't have permission to edit the column... so we 
				// need to cut off access right now!!!
				error_log("Insufficient permissions to access field $field[name] by current user.");
				throw new Exception("Failed to get options for field $filed[name] because you don't have view permission for this field.", 401);
			
			}
			
			
			
			
			if ( !@$field['widget']['table'] ){
				error_log("widget:table not defined for field ".$field['name']." of table ".$table->tablename.".");
				throw new Exception(self::errorMessage($fieldname), 500);
			}
			$targetTable = Dataface_Table::loadTable($field['widget']['table']);
			
			$filters = array();
			if ( @$field['widget']['filters']  and is_array($field['widget']['filters'])){
				foreach ($field['widget']['filters'] as $key=>$val){
					if ( isset($query[$key]) ){
						$filters[$key]=$query[$key];
					} else if ( strpos($val,'$') === 0 ){
						$filters[$key] = '=';
					} else {
						$filters[$key] = $val;
					}
				}
			}
			
			
			
			$limit = 250;
			if ( @$field['widget']['limit'] ){
				$limit = intval($field['widget']['limit']);
			}
			
			$filters['-limit'] = $limit;
			$records = df_get_records_array($field['widget']['table'], $filters);
			//if ( count($filters) > 1 ){
			//	echo "Num records found: ".count($records);
			//	print_r($filters);
			//}
			$keyCol = null;
			$labelCol = null;
			
			if ( @$field['widget']['keycol'] ){
				$keyCol = $field['widget']['keycol'];
			} else {
				foreach ($targetTable->keys() as $k=>$v){
					$keyCol = $k;
					break;
				}
			}
			
			if ( @$field['widget']['labelcol'] ){
				$labelCol = $field['widget']['labelcol'];
			}
			
			$out = array();
			foreach ($records as $r){
				if ( @$field['widget']['ignore_permissions'] ){
					$r->secureDisplay = false;
				} else {
					if ( !$r->checkPermission('view') ) continue;
					if ( !$r->checkPermission('view', array('field'=>$keyCol))) continue;
					if ( $labelCol and !$r->checkPermission('view', array('field'=>$labelCol))) continue;
				}
				
				if ( $labelCol ){
					$out[$r->val($keyCol)] = $r->display($labelCol);
				} else {
					$out[$r->val($keyCol)] = $r->getTitle();
				}
			
			}
			
			$this->out(array(
				'code'=>200,
				'message'=>'Received default valuelist',
				'values'=>$out
			));
			exit;
			
		
		} catch (Exception $ex){
		
			$this->out(array(
				'code'=>$ex->getCode(),
				'message'=>$ex->getMessage()
			));
			exit;
		}
	
	}
	
	
	/**
	 * @brief Outputs content as JSON.  Outputs headers as necessary.
	 */
	private function out($params){
		header('Content-type: text/json; charset="'.Dataface_Application::getInstance()->_conf['oe'].'"');
		echo json_encode($params);
	}
	
	
	
}